lib/logstash/filters/csv.rb

```diff
@@ -151,10 +151,14 @@ def filter(event)
   private
 
   # construct the correct Event field reference for given field_name, taking into account @target
-  # @param field_name [String] the bare field name without brakets
+  # @param field_name [String] the field name.
   # @return [String] fully qualified Event field reference also taking into account @target prefix
   def field_ref(field_name)
-    "#{@target}[#{field_name}]"
+    if field_name.start_with?("[")
```

Comment:
I'm confused here. The `field_name` is the name of a column a user has chosen. Would the `true` side of the if only occur if someone defined a field that starts with a bracket? Do I understand the problem here correctly?

Reply:
We're essentially doing the same thing slightly differently in multiple places, which makes it a little hard to see what is going on.

I think that the functionality could be made a little more clear if we were to provide a couple smaller methods that could be _composed_ to produce the same effect: one to take either a field name or a bracketed field reference and convert it to a field reference (e.g., `foo` becomes `[foo]`, `[foo][bar]` remains `[foo][bar]`), and another one to prepend the `@target`:

~~~ diff
diff --git a/lib/logstash/filters/csv.rb b/lib/logstash/filters/csv.rb
index 51ca33f..a25ae03 100644
--- a/lib/logstash/filters/csv.rb
+++ b/lib/logstash/filters/csv.rb
@@ -104,8 +104,8 @@ class LogStash::Filters::CSV < LogStash::Filters::Base
     # @convert_symbols contains the symbolized types to avoid symbol conversion in the transform method
     @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
 
-    # make sure @target is in the format [field name] if defined, i.e. surrounded by brakets
-    @target = "[#{@target}]" if @target && @target !~ /^\[[^\[\]]+\]$/
+    # make sure @target is in the format [field name] if defined, i.e. surrounded by brackets
+    @target = @target && field_ref_fragment(@target)
     
     # if the zero byte character is entered in the config, set the value
     if (@quote_char == "\\x00")
@@ -154,7 +154,19 @@ class LogStash::Filters::CSV < LogStash::Filters::Base
   # @param field_name [String] the bare field name without brakets
   # @return [String] fully qualified Event field reference also taking into account @target prefix
   def field_ref(field_name)
-    "#{@target}[#{field_name}]"
+    "#{@target}#{field_ref_fragment(field_name)}"
+  end
+
+  # construct a valid bracket-syntax field reference fragment from the given field name or reference.
+  # if nil is provided, nil will be produced.
+  #
+  # @param field_name_or_ref [String] a no-bracket field name (e.g., 'foo') or
+  #                                   a bracket-syntax field reference (e.g., '[foo][bar]')
+  # @return [String] field reference fragment in bracket syntax (e.g., '[foo]')
+  def field_ref_fragment(field_name_or_ref)
+    return field_name_or_ref if field_name_or_ref.nil? || field_name_or_ref.start_with?('[')
+
+    "[#{field_name_or_ref}]"
   end
 
   def ignore_field?(index)
~~~

Reply:
+1 on refactoring to make this simpler. Composing fieldreferences could be done more easily if https://github.com/elastic/logstash/issues/8307 is implemented.